Mathlib/NumberTheory/LSeries/Convergence.lean

```diff
@@ -0,0 +1,96 @@
+when `re s < x`. -/
+noncomputable def abscissaOfAbsConv (f : ArithmeticFunction ℂ) : EReal :=
+  sInf <| Real.toEReal '' {x : ℝ | LSeriesSummable f x}
+
+/-- An open right half plane is open. -/
+lemma _root_.Complex.isOpen_rightHalfPlane (x : EReal) : IsOpen {z : ℂ | x < z.re} :=
+  isOpen_lt continuous_const <| EReal.continuous_coe_iff.mpr continuous_re
+
+lemma LSeriesSummable_of_abscissaOfAbsConv_lt_re {f : ArithmeticFunction ℂ} {s : ℂ}
+    (hs : abscissaOfAbsConv f < s.re) : LSeriesSummable f s := by
+  simp only [abscissaOfAbsConv, sInf_lt_iff, Set.mem_image, Set.mem_setOf_eq,
+    exists_exists_and_eq_and, EReal.coe_lt_coe_iff] at hs
+  obtain ⟨y, hy, hys⟩ := hs
+  exact hy.of_re_le_re <| ofReal_re y ▸ hys.le
+
+lemma LSeriesSummable_lt_re_of_abscissaOfAbsConv_lt_re {f : ArithmeticFunction ℂ} {s : ℂ}
+    (hs : abscissaOfAbsConv f < s.re) :
+    ∃ x : ℝ, x < s.re ∧ LSeriesSummable f x := by
+  obtain ⟨x, hx₁, hx₂⟩ := EReal.exists_between_coe_real hs
+  refine ⟨x, EReal.coe_lt_coe_iff.mp hx₂, LSeriesSummable_of_abscissaOfAbsConv_lt_re hx₁⟩
+
+lemma LSeriesSummable.abscissaOfAbsConv_le {f : ArithmeticFunction ℂ} {s : ℂ}
+    (h : LSeriesSummable f s) : abscissaOfAbsConv f ≤ s.re := by
+  refine sInf_le <| Membership.mem.out ?_
+  simp only [Set.mem_setOf_eq, Set.mem_image, EReal.coe_eq_coe_iff, exists_eq_right]
+  exact h.of_re_le_re <| by simp only [ofReal_re, le_refl]
+
+lemma abscissaOfAbsConv_le_of_forall_lt_LSeriesSummable {f : ArithmeticFunction ℂ} {x : ℝ}
+    (h : ∀ y : ℝ, x < y → LSeriesSummable f y) :
+    abscissaOfAbsConv f ≤ x := by
+  refine sInf_le_iff.mpr fun y hy ↦ ?_
+  simp only [mem_lowerBounds, Set.mem_image, Set.mem_setOf_eq, forall_exists_index, and_imp,
+    forall_apply_eq_imp_iff₂] at hy
+  have H (a : EReal) : x < a → y ≤ a := by
+    induction' a using EReal.rec with a₀
+    · simp only [not_lt_bot, le_bot_iff, IsEmpty.forall_iff]
+    · exact_mod_cast fun ha ↦ hy a₀ (h a₀ ha)
+    · simp only [EReal.coe_lt_top, le_top, forall_true_left]
+  exact Set.Ioi_subset_Ici_iff.mp H
+
+lemma abscissaOfAbsConv_le_of_forall_lt_LSeriesSummable' {f : ArithmeticFunction ℂ} {x : EReal}
+    (h : ∀ y : ℝ, x < y → LSeriesSummable f y) :
+    abscissaOfAbsConv f ≤ x := by
+  induction' x using EReal.rec with y
+  · refine le_of_eq <| sInf_eq_bot.mpr fun y hy ↦ ?_
+    induction' y using EReal.rec with z
+    · simp only [gt_iff_lt, lt_self_iff_false] at hy
+    · exact ⟨z - 1,  ⟨z-1, h (z - 1) <| EReal.bot_lt_coe _, rfl⟩, by norm_cast; exact sub_one_lt z⟩
+    · exact ⟨0, ⟨0, h 0 <| EReal.bot_lt_coe 0, rfl⟩, EReal.zero_lt_top⟩
+  · exact abscissaOfAbsConv_le_of_forall_lt_LSeriesSummable <| by exact_mod_cast h
+  · exact le_top
+
+/-- If `‖f n‖` is bounded by a constant times `n^x`, then the abscissa of absolute convergence
+of `f` is bounded by `x + 1`. -/
+lemma abscissaOfAbsConv_le_of_le_const_mul_rpow {f : ArithmeticFunction ℂ} {x : ℝ}
+    (h : ∃ C, ∀ n, ‖f n‖ ≤ C * n ^ x) : abscissaOfAbsConv f ≤ x + 1 := by
```

Comment:
Maybe this lemma and the next might be better formulated as an "IsBigO" statement?

Reply:
I'll think about that.

Reply:
To be precise I meant that the hypothesis `h` could become `f =O[atTop] fun n => n ^ x`, which is almost definitionally equal to what you wrote but has a richer API attached to it – I'm not suggesting any change to the conclusion of the lemma.

Reply:
I have now added a "isBigO" version of `LSeriesSummable_of_le_const_mul_rpow`, which did require some additional work to deal with the fact that O() means that the bound holds eventually and not a priori everywhere, and then used that to provide "isBigO" versions of these two lemmas (in addition, as the other versions might still be useful).